tests/test_coef.py

```diff
@@ -0,0 +1,369 @@
+def img_size(request):
+    """
+    Image size for basis.
+    """
+    return request.param
+
+
+@pytest.fixture(params=STACKS, ids=lambda x: f"stack={x}", scope="module")
+def stack(request):
+    """
+    Stack dimensions.
+    """
+    return request.param
+
+
+@pytest.fixture(params=ALLYOURBASES, ids=lambda x: f"basis={x}", scope="module")
+def basis(request, img_size, basis_dtype):
+    """
+    Parameterized `Basis` instantiation.
+    """
+    cls = request.param
+    return cls(img_size, dtype=basis_dtype)
+
+
+@pytest.fixture(scope="module")
+def coef_fixture(basis, stack, dtype):
+    """
+    Construct parameterized testing coefficient array as `Coef`.
+    """
+    # Combine the stack and coefficent counts into multidimensional
+    # shape.
+    size = stack + (basis.count,)
+
+    coef_np = np.random.random(size=size).astype(dtype, copy=False)
+
+    return Coef(basis, coef_np, dtype=dtype)
+
+
+def test_mismatch_count(basis):
+    """
+    Confirm raises when instantiated with incorrect coef vector len.
+    """
+    # Derive an incorrect Coef
+    x = np.empty(basis.count + 1, basis.dtype)
+    with pytest.raises(RuntimeError, match=r".*does not match basis count.*"):
+        _ = Coef(basis, x)
+
+
+def test_incorrect_coef_type(basis):
+    """
+    Confirm raises when instantiated with incorrect coef type.
+    """
+    # Construct incorrect Coef type (list)
+    x = list(range(basis.count + 1))
+    with pytest.raises(ValueError, match=r".*should be instantiated with an.*"):
+        _ = Coef(basis, x)
+
+
+def test_0dim(basis):
+    """
+    Confirm raises when instantiated with 0dim scalar.
+    """
+    # Construct 0dim scalar object
+    x = np.array(1)
+    with pytest.raises(ValueError, match=r".*with shape.*"):
+        _ = Coef(basis, x)
+
+
+def test_not_a_basis():
+    """
+    Confirm raises when instantiated with something that is not a Basis.
+    """
+    # Derive an incorrect Coef
+    x = np.empty(10)
+    with pytest.raises(TypeError, match=r".*required to be a `Basis`.*"):
+        _ = Coef(None, x)
+
+
+def test_coef_key_dims(coef_fixture):
+    """
+    Test key lookup out of bounds dimension raises.
+    """
+    dim = coef_fixture.ndim
+    # Construct a key with too many dims
+    key = (0,) * (dim + 1)
+    with pytest.raises(ValueError, match=r".*stack_dim is.*"):
+        _ = coef_fixture[key]
+
+
+def test_incorrect_reshape(basis):
+    """
+    Confirm raises when attempting incorrect stack reshape.
+    """
+
+    # create a multi dim coef array.
+    x = np.empty((2, 3, 4, basis.count))
+    c = Coef(basis, x)
+
+    # Alter the stack shape, creating an incorrect shape.
+    shp = list(c.stack_shape)
+    shp[0] = shp[0] + 1
+
+    with pytest.raises(ValueError, match=r".*cannot be reshaped to.*"):
+        _ = c.stack_reshape(*shp)
+
+
+def test_stack_reshape(basis):
+    """
+    Test stack_reshape matches corresponding pure Numpy reshape.
```

Comment:
Is flatten the only relevant stack reshape to test? Should we test a permutation of the stack shape?

Reply:
Are you looking at old code?

Reply:
Takesies backsies.  This code already tests that case. I spelt it out in the comments.

Reply:
[c30387a](https://github.com/ComputationalCryoEM/ASPIRE-Python/pull/1015/commits/c30387aa54a78b80c65f1f81ca6623df7dff716a)

Reply:
Ooops, read "flatten" and `prod` and my eyes deceived me. My bad.